tests/test/unsize.rs

```diff
@@ -586,3 +586,200 @@ fn struct_unsizing() {
         }
     }
 }
+
+#[test]
+fn dispatch_from_dyn() {
+    test! {
+        program {
+            #[lang(dispatch_from_dyn)]
+            trait DispatchFromDyn<T> {}
+
+            impl<'a, T, U> DispatchFromDyn<&'a U> for &'a T {}
+        }
+
+        // Smoke test that DispatchFromDyn works just like any other impl.
+        goal {
+            forall<'a> {
+                &'a u8: DispatchFromDyn<&'a u8>
+            }
+        } yields {
+            "Unique"
+        }
+    }
+}
+
+#[test]
+fn dispatch_from_dyn_wf() {
+    lowering_success! {
+        program {
+            #[lang(dispatch_from_dyn)]
+            trait DispatchFromDyn<T> {}
+
+            #[one_zst]
+            struct PhantomData<T> {}
```

Comment:
This should probably be named something other than `PhantomData` as to not be confusing.